src/classes/AutoNumberService.cls

```diff
@@ -0,0 +1,326 @@
+                    WHERE Object_API_Name__c = :String.valueOf(sObjType)
+                    AND IsActive__c = true
+            ];
+        }
+    }
+
+    // TDTM expects local names without namespaces for Class and Object
+    private final String classNameForTDTM = 'AutoNumber_TDTM';
+    @TestVisible
+    Boolean isTriggerHandlerEnabled {
+        get {
+            return ![
+                    SELECT Id
+                    FROM Trigger_Handler__c
+                    WHERE Object__c = :sObjType.getDescribe().getLocalName()
+                    AND Class__c = :classNameForTDTM
+                    AND Active__c = true
+            ].isEmpty();
+        }
+    }
+
+    @TestVisible
+    Trigger_Handler__c triggerHandler {
+        get {
+            return new Trigger_Handler__c(
+                    Active__c = true,
+                    Asynchronous__c = false,
+                    Class__c = classNameForTDTM,
+                    Load_Order__c = 0,
+                    Object__c = sObjType.getDescribe().getLocalName(),
+                    Trigger_Action__c = 'AfterInsert;');
+        }
+    }
+
+    /**
+     * @description Wrapper class used to provide easier access to calculated
+     * properties for AutoNumber__c records.
+     */
+    private class AutoNumber {
+        AutoNumber__c record;
+        Integer nextAutoNumber;
+        AutoNumberService autoNumberService;
+        private SObjectType sObjType {
+            get {
+                return UTIL_Describe.getSObjectType(record.Object_API_Name__c);
+            }
+        }
+        private SObjectField sObjField {
+            get {
+                return UTIL_Describe.getFieldDescribe(String.valueOf(sObjType),
+                        record.Field_API_Name__c).getSobjectField();
+            }
+        }
+        private String displayFormat {
+            get {
+                return record.Display_Format__c;
+            }
+        }
+
+        public AutoNumber(String autoNumberJSON, AutoNumberService ans) {
+            this.record = convertJSONToAutoNumber(autoNumberJSON);
+            this.autoNumberService = ans;
+            this.validate();
+        }
+
+        // Used in trigger context by apply()
+        public AutoNumber(AutoNumber__c record) {
+            this.record = record;
+            this.nextAutoNumber = getNextAutoNumber();
+        }
+
+        public String getFormattedNumber(Integer num) {
+            return String.format(displayFormat,
+                    new List<Object>{
+                            String.valueOf(num).leftPad(getNumberLength(), '0')
+                    }
+            );
+        }
+
+        public AutoNumber__c save() {
+            insert record;
+            return record;
+        }
+
+        public void setMaxUsedNumber(Integer num) {
+            record.Max_Used_Number__c = num;
+            update record;
+        }
+
+        private AutoNumber__c convertJSONToAutoNumber(String autoNumber) {
```

Comment:
Out of curiosity, are we setting the fields one by one because we can't deserialize directly due to the protected nature of AutoNumber__c?

Reply:
Yep - deserializing to the AutoNumber__c class throws an insufficient access exception.

![2020-07-28_17-30-00](https://user-images.githubusercontent.com/46458081/88743023-87617780-d0f8-11ea-83c7-24a669f8040a.png)